File: src/target26_1/java/com/skysoft/mixin/BufferBuilderMixin.java
package com.skysoft.mixin;

import com.mojang.blaze3d.vertex.BufferBuilder;
import com.mojang.blaze3d.vertex.VertexFormatElement;
import com.skysoft.utils.render.shader.VertexMemoryAccess;
import org.spongepowered.asm.mixin.Mixin;
import org.spongepowered.asm.mixin.Shadow;

@Mixin(BufferBuilder.class)
public abstract class BufferBuilderMixin implements VertexMemoryAccess {
    @Shadow
    protected abstract long beginElement(VertexFormatElement element);

    @Override
    public long skysoft$attributeAddress(VertexFormatElement element, int byteOffset) {
        return beginElement(element);
    }
}

File: src/target26_1/kotlin/com/skysoft/mixin/BufferBuilderMixin.kt
package com.skysoft.mixin

import com.mojang.blaze3d.vertex.BufferBuilder
import com.mojang.blaze3d.vertex.VertexFormatElement
import com.skysoft.utils.render.shader.VertexMemoryAccess
import org.spongepowered.asm.mixin.Mixin
import org.spongepowered.asm.mixin.Shadow

@Mixin(BufferBuilder::class)
abstract class BufferBuilderMixin : VertexMemoryAccess {
    @Shadow
    protected abstract fun beginElement(element: VertexFormatElement): Long

    override fun skysoftAttributeAddress(element: VertexFormatElement, byteOffset: Int): Long =
        beginElement(element)
}
